app/screens/find_channels/filtered_list/filtered_list.tsx

```diff
@@ -1,44 +1,176 @@
+export const MAX_RESULTS = 20;
+
+const sortyByUserOrChannel = <T extends Channel |UserModel>(locale: string, teammateDisplayNameSetting: string, a: T, b: T): number => {
+    const aDisplayName = 'display_name' in a ? a.display_name : displayUsername(a, locale, teammateDisplayNameSetting);
+    const bDisplayName = 'display_name' in b ? b.display_name : displayUsername(b, locale, teammateDisplayNameSetting);
+
+    return aDisplayName.toLowerCase().localeCompare(bDisplayName.toLowerCase(), locale, {numeric: true});
+};
+
+const FilteredList = ({
+    archivedChannels, close, channelsMatch, channelsMatchStart, currentTeamId,
+    keyboardHeight, loading, onLoading, restrictDirectMessage, showTeamName,
+    teamIds, teammateDisplayNameSetting, term, usersMatch, usersMatchStart,
+}: Props) => {
+    const bounce = useRef<DebouncedFunc<() => void>>();
+    const mounted = useRef(false);
     const serverUrl = useServerUrl();
-    const [data] = useState([]);
+    const theme = useTheme();
+    const {locale, formatMessage} = useIntl();
     const flatListStyle = useMemo(() => ({flexGrow: 1, paddingBottom: keyboardHeight}), [keyboardHeight]);
+    const [remoteChannels, setRemoteChannels] = useState<RemoteChannels>({archived: [], startWith: [], matches: []});
+    const totalLocalResults = channelsMatchStart.length + channelsMatch.length + usersMatchStart.length;
+
+    const search = async () => {
+        onLoading(true);
+        if (mounted.current) {
+            setRemoteChannels({archived: [], startWith: [], matches: []});
+        }
+        const lowerCasedTerm = (term.startsWith('@') ? term.substring(1) : term).toLowerCase();
+        if ((channelsMatchStart.length + channelsMatch.length) < MAX_RESULTS) {
+            if (restrictDirectMessage) {
+                searchProfiles(serverUrl, lowerCasedTerm, {team_id: currentTeamId, allow_inactive: true});
+            } else {
+                searchProfiles(serverUrl, lowerCasedTerm, {allow_inactive: true});
+            }
+        }
+
+        if (!term.startsWith('@')) {
+            if (totalLocalResults < MAX_RESULTS) {
+                const {channels} = await searchAllChannels(serverUrl, lowerCasedTerm, true);
+                if (channels) {
+                    const existingChannelIds = new Set(channelsMatchStart.concat(channelsMatch).concat(archivedChannels).map((c) => c.id));
+                    const filteredChannels = channels.filter((c) => !existingChannelIds.has(c.id) && teamIds.has(c.team_id));
+                    const startWith = filteredChannels.filter((c) => c.display_name.toLowerCase().startsWith(term) && c.delete_at === 0).
+                        sort((a, b) => (a.last_post_at > b.last_post_at ? 1 : -1)).slice(0, MAX_RESULTS + 1);
+                    const matches = filteredChannels.filter((c) => c.display_name.toLowerCase().includes(term) && !c.display_name.toLowerCase().startsWith(term) && c.delete_at === 0).
```

Comment:
I wonder if we it makes sense to move all this to a reduce like this one. O(n) === O(4n), but still, feels better.

```
const [startWith, matches, archived] = channels.reduce<[Channel[], Channel[], Channel[]]>(([sw, m, a], c) => {
    if (existingChannelIds.has(c.id) {
        return [sw, m, a];
    }
    if (c.delete_at === 0) {
        if (c.display_name.toLowerCase().startsWith(term)) {
            return [[...sw, c], m, a];
        }
        if (c.display_name.toLowerCase().includes(term)) {
            return [sw, [...m, c], a];
        }
        return [sw, m, a];
    }
    if (c.display_name.toLowerCase().includes(term)) {
         return [sw, m, [...a, c];
    }
    return [sw, m, c];
}
```

Reply:
interesting.. yes, I think something like this should work